# Background

Pada kesempatan kali ini kita akan membuat sebuah dashboard menggunakan dash. Data yang digunakan adalah data yang bersumber
dari [Kaggle] https://www.kaggle.com/datasets/pavansubhasht/ibm-hr-analytics-attrition-dataset).Data tersebut berisi tentang data attrition (pengurangan) karyawan yang terjadi di perusahaan IBM. 

# Read Data & Library

In [43]:
# import library
import pandas as pd

In [44]:
# read data
attrition = pd.read_csv('WA_Fn-UseC_-HR-Employee-Attrition.csv')
attrition.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1            80                 0   
1  ...                         4            80                 1   
2  ...                         2            80                 0   
3  ...                         3            80                 0   
4  ...                         4            80                 1   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                      0               1               6   
1                 10                      3               3              10   
2                  7                      3               3               0   
3                  8                      3               3               8   
4                  6                      3               3               2   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                     5  
1                  7                        1                     7  
2                  0                        0                     0  
3                  7                        3                     0  
4                  2                        2                     2  

[5 rows x 35 columns]

In [45]:
# cek jumlah baris dan kolom
attrition.shape

(1470, 35)

**Keterangan** \
Dari data diatas didapatkan bahwa data yang kita miliki terdiri dari 1470 baris dan 35 kolom

In [46]:
# cek tipe data awal
attrition.dtypes

Age                          int64
Attrition                   object
BusinessTravel              object
DailyRate                    int64
Department                  object
DistanceFromHome             int64
Education                    int64
EducationField              object
EmployeeCount                int64
EmployeeNumber               int64
EnvironmentSatisfaction      int64
Gender                      object
HourlyRate                   int64
JobInvolvement               int64
JobLevel                     int64
JobRole                     object
JobSatisfaction              int64
MaritalStatus               object
MonthlyIncome                int64
MonthlyRate                  int64
NumCompaniesWorked           int64
Over18                      object
OverTime                    object
PercentSalaryHike            int64
PerformanceRating            int64
RelationshipSatisfaction     int64
StandardHours                int64
StockOptionLevel             int64
TotalWorkingYears   

**Informasi Kolom :**

- Age : Usia dari masing-masing pekerja
- Attrition : Status dari attrition pekerja 
- BusinessTravel :
- DailyRate : Upah harian
- Department : Departemen yang ada di perusahaan
- DistanceFromHome : Jarak tempat kerja dari rumah
- Education : Pendidikan 
- EducationField : Jurusan
- EmployeeCount : Jumlah pegawai 
- EmployeeNumber : Nomor dari masing-masing pekerja
- EnvironmentSatisfaction : Tingkat kepuasan pekerja
- Gender : Jenis Kelamin
- HourlyRate : Tarif perjam
- JobInvolvement : Keterlibatan pekerja dalam suatu pekerjaan
- JobLevel : Tingkat pekerjaan
- JobRole : Peran dalam pekerjaan
- JobSatisfaction : Kepuasan dalam bekerja
- MaritalStatus : Status perkawinan
- MonthlyIncome : Penghasilan perbulan
- MonthlyRate : Tarif perbulan
- NumCompaniesWorked : Jumlah perusahaan bekerja
- Over18 : Pekerja diatas 18 tahun
- OverTime : Lembur
- PercentSalaryHike : Persentase kenaikan gaji
- PerformanceRating : Peringkat kinerja
- RelationshipSatisfaction : Kepuasan hubungan dalam bekerja
- StandardHours : Standar jam kerja
- StockOptionLevel : Tingkat opsi saham
- TotalWorkingYears : Total tahun bekerja
- TrainingTimesLastYear : Waktu pelatihan tahun terakhir
- WorkLifeBalance : Keseimbangan kehidupan kerja
- YearsAtCompany : Lama bekerja diperusahaan dalam tahun
- YearsInCurrentRole : Tahun dalam peran saat ini 
- YearsSinceLastPromotion : Tahun sejak promosi terakhir
- YearsWithCurrManager : Lamanya bekerja sama dalam tahun dengan Manajer yang sekarang

# Data Preprocessing 

In [47]:
# replace kolom Education
attrition['Education'] = attrition['Education'].replace(1, 'Below College')
attrition['Education'] = attrition['Education'].replace(2, 'College')
attrition['Education'] = attrition['Education'].replace(3, 'Bachelor')
attrition['Education'] = attrition['Education'].replace(4, 'Master')
attrition['Education'] = attrition['Education'].replace(5, 'Doctor')

In [48]:
# replace kolom EnvironmentSatisfaction
attrition['EnvironmentSatisfaction'] = attrition['EnvironmentSatisfaction'].replace(1 , 'Low')
attrition['EnvironmentSatisfaction'] = attrition['EnvironmentSatisfaction'].replace(2 , 'Medium')
attrition['EnvironmentSatisfaction'] = attrition['EnvironmentSatisfaction'].replace(3 , 'High')
attrition['EnvironmentSatisfaction'] = attrition['EnvironmentSatisfaction'].replace(4 , 'Very High')

In [49]:
# replace kolom JobInvolvement
attrition['JobInvolvement'] = attrition['JobInvolvement'].replace(1 , 'Low')
attrition['JobInvolvement'] = attrition['JobInvolvement'].replace(2 , 'Medium')
attrition['JobInvolvement'] = attrition['JobInvolvement'].replace(3 , 'High')
attrition['JobInvolvement'] = attrition['JobInvolvement'].replace(4 , 'Very High')

In [50]:
# replace kolom JobSatisfaction
attrition['JobSatisfaction'] = attrition['JobSatisfaction'].replace(1 , 'Low')
attrition['JobSatisfaction'] = attrition['JobSatisfaction'].replace(2 , 'Medium')
attrition['JobSatisfaction'] = attrition['JobSatisfaction'].replace(3 , 'High')
attrition['JobSatisfaction'] = attrition['JobSatisfaction'].replace(4 , 'Very High')

In [51]:
# replace kolom PerformanceRating
attrition['PerformanceRating'] = attrition['PerformanceRating'].replace(1 , 'Low')
attrition['PerformanceRating'] = attrition['PerformanceRating'].replace(2 , 'Good')
attrition['PerformanceRating'] = attrition['PerformanceRating'].replace(3 , 'Excellent')
attrition['PerformanceRating'] = attrition['PerformanceRating'].replace(4 , 'Outstanding')

In [52]:
# replace kolom RelationshipSatisfaction
attrition['RelationshipSatisfaction'] = attrition['RelationshipSatisfaction'].replace(1 , 'Low')
attrition['RelationshipSatisfaction'] = attrition['RelationshipSatisfaction'].replace(2 , 'Medium')
attrition['RelationshipSatisfaction'] = attrition['RelationshipSatisfaction'].replace(3 , 'High')
attrition['RelationshipSatisfaction'] = attrition['RelationshipSatisfaction'].replace(4 , 'Very High')

In [53]:
# replace kolom WorkLifeBalance
attrition['WorkLifeBalance'] = attrition['WorkLifeBalance'].replace(1 , 'Bad')
attrition['WorkLifeBalance'] = attrition['WorkLifeBalance'].replace(2 , 'Good')
attrition['WorkLifeBalance'] = attrition['WorkLifeBalance'].replace(3 , 'Better')
attrition['WorkLifeBalance'] = attrition['WorkLifeBalance'].replace(4 , 'Best')

In [54]:
# sesuaikan tipe data
attrition['Attrition'] = attrition['Attrition'].astype('category')
attrition['BusinessTravel'] = attrition['BusinessTravel'].astype('category')
attrition['Department'] = attrition['Department'].astype('category')
attrition['Education'] = attrition['Education'].astype('category')
attrition['EducationField'] = attrition['EducationField'].astype('category')
attrition['EnvironmentSatisfaction'] = attrition['EnvironmentSatisfaction'].astype('category')
attrition['Gender'] = attrition['Gender'].astype('category')
attrition['JobInvolvement'] = attrition['JobInvolvement'].astype('category')
attrition['JobRole'] = attrition['JobRole'].astype('category')
attrition['JobSatisfaction'] = attrition['JobSatisfaction'].astype('category')
attrition['MaritalStatus'] = attrition['MaritalStatus'].astype('category')
attrition['OverTime'] = attrition['OverTime'].astype('category')
attrition['PerformanceRating'] = attrition['PerformanceRating'].astype('category')
attrition['RelationshipSatisfaction'] = attrition['RelationshipSatisfaction'].astype('category')
attrition['WorkLifeBalance'] = attrition['WorkLifeBalance'].astype('category')

In [55]:
# cek tipe data lagi apakah sudah sesuai
attrition.dtypes

Age                            int64
Attrition                   category
BusinessTravel              category
DailyRate                      int64
Department                  category
DistanceFromHome               int64
Education                   category
EducationField              category
EmployeeCount                  int64
EmployeeNumber                 int64
EnvironmentSatisfaction     category
Gender                      category
HourlyRate                     int64
JobInvolvement              category
JobLevel                       int64
JobRole                     category
JobSatisfaction             category
MaritalStatus               category
MonthlyIncome                  int64
MonthlyRate                    int64
NumCompaniesWorked             int64
Over18                        object
OverTime                    category
PercentSalaryHike              int64
PerformanceRating           category
RelationshipSatisfaction    category
StandardHours                  int64
S

**Keterangan**\
Tipe data sudah sesuai  

In [56]:
# cek apakah ada data yang kosong
attrition.isna().sum()

Age                         0
Attrition                   0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EmployeeNumber              0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
MonthlyIncome               0
MonthlyRate                 0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
YearsInCurrentRole          0
YearsSince

**Keterangan**\
Ternyata dari data yang kita miliki tidak terdapat missing value atau data yang hilang

In [57]:
# cek 5 data teratas lagi
attrition.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome      Education EducationField  EmployeeCount  \
0                 1        College  Life Sciences              1   
1                 8  Below College  Life Sciences              1   
2                 2        College          Other              1   
3                 3         Master  Life Sciences              1   
4                 2  Below College        Medical              1   

   EmployeeNumber  ... RelationshipSatisfaction StandardHours  \
0               1  ...                      Low            80   
1               2  ...                Very High            80   
2               4  ...                   Medium            80   
3               5  ...                     High            80   
4               7  ...                Very High            80   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  \
0                 0                 8                      0             Bad   
1                 1                10                      3          Better   
2                 0                 7                      3          Better   
3                 0                 8                      3          Better   
4                 1                 6                      3          Better   

  YearsAtCompany YearsInCurrentRole  YearsSinceLastPromotion  \
0              6                  4                        0   
1             10                  7                        1   
2              0                  0                        0   
3              8                  7                        3   
4              2                  2                        2   

   YearsWithCurrManager  
0                     5  
1                     7  
2                     0  
3                     0  
4                     2  

[5 rows x 35 columns]

# Visualization

## Import Library

In [58]:
import plotly as pl
import plotly.express as px

## Business Questions

**Pertanyaan Nomor 1**\
1.Berapa jumlah pekerja yang terdapat dalam data yang kita miliki?

In [59]:
# banyaknya jumlah pekerja yang ada di dalam data kita
attrition.shape[0]

1470

**Pertanyaan Nomor 2**\
2. Berapa banyak pekerja yang Attrition Yes dan Attrition No?

In [60]:
# attrition yes
attrition[attrition['Attrition'] == 'Yes'].shape[0]

237

In [61]:
# attrition no
attrition[attrition['Attrition'] == 'No'].shape[0]

1233

**Pertanyaan Nomor 3**\
3. Dari data yang kita miliki, gender manakah yang paling banyak **Attrition** Yes? Buat visualisasinya

In [62]:
# subset attrition yes
attrition_yes = attrition[attrition['Attrition'] == 'Yes']
attrition_yes

Age Attrition     BusinessTravel  DailyRate              Department  \
0      41       Yes      Travel_Rarely       1102                   Sales   
2      37       Yes      Travel_Rarely       1373  Research & Development   
14     28       Yes      Travel_Rarely        103  Research & Development   
21     36       Yes      Travel_Rarely       1218                   Sales   
24     34       Yes      Travel_Rarely        699  Research & Development   
...   ...       ...                ...        ...                     ...   
1438   23       Yes  Travel_Frequently        638                   Sales   
1442   29       Yes      Travel_Rarely       1092  Research & Development   
1444   56       Yes      Travel_Rarely        310  Research & Development   
1452   50       Yes  Travel_Frequently        878                   Sales   
1461   50       Yes      Travel_Rarely        410                   Sales   

      DistanceFromHome      Education    EducationField  EmployeeCount  \
0                    1        College     Life Sciences              1   
2                    2        College             Other              1   
14                  24       Bachelor     Life Sciences              1   
21                   9         Master     Life Sciences              1   
24                   6  Below College           Medical              1   
...                ...            ...               ...            ...   
1438                 9       Bachelor         Marketing              1   
1442                 1         Master           Medical              1   
1444                 7        College  Technical Degree              1   
1452                 1         Master     Life Sciences              1   
1461                28       Bachelor         Marketing              1   

      EmployeeNumber  ... RelationshipSatisfaction StandardHours  \
0                  1  ...                      Low            80   
2                  4  ...                   Medium            80   
14                19  ...                   Medium            80   
21                27  ...                   Medium            80   
24                31  ...                     High            80   
...              ...  ...                      ...           ...   
1438            2023  ...                      Low            80   
1442            2027  ...                   Medium            80   
1444            2032  ...                Very High            80   
1452            2044  ...                Very High            80   
1461            2055  ...                   Medium            80   

      StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  \
0                    0                 8                      0   
2                    0                 7                      3   
14                   0                 6                      4   
21                   0                10                      4   
24                   0                 8                      2   
...                ...               ...                    ...   
1438                 1                 1                      3   
1442                 3                 4                      3   
1444                 1                14                      4   
1452                 2                12                      3   
1461                 1                20                      3   

     WorkLifeBalance YearsAtCompany YearsInCurrentRole  \
0                Bad              6                  4   
2             Better              0                  0   
14            Better              4                  2   
21            Better              5                  3   
24            Better              4                  2   
...              ...            ...                ...   
1438            Good              1                  0   
1442            Best              2                  2   
1444             Bad             10             

In [63]:
# lihat jumlah frekuensinya
attrition_yes['Gender'].value_counts()

Male      150
Female     87
Name: Gender, dtype: int64

In [64]:
# visualisasi
attrition_yes_gender = attrition_yes['Gender'].value_counts()
px.bar(attrition_yes_gender,
      labels = {
          'index' : 'Gender',
          'value' : 'Jumlah Karyawan'
      },
      title='Gender Manakah Yang Paling Banyak Attrition Yes?',
      template='presentation',
      color_discrete_sequence=['#F5F5DC'],
      )

**Pertanyaan Nomor 4**\
4. Berapa banyak attrition dari tiap pekerja disetiap departemen ? Buat visualisasinya  

In [65]:
attrition_agg = attrition.groupby(['Department','Attrition']).count()[['EmployeeNumber']].reset_index()
attrition_agg

Department Attrition  EmployeeNumber
0         Human Resources        No              51
1         Human Resources       Yes              12
2  Research & Development        No             828
3  Research & Development       Yes             133
4                   Sales        No             354
5                   Sales       Yes              92

In [66]:
# visualisasikan
px.bar(attrition_agg.sort_values('EmployeeNumber'),
       x='EmployeeNumber',
       y='Department',
       color='Attrition',
       orientation='h',
       barmode = 'group',
      )

**Pertanyaan Nomor 5**\
5. Bagaimana jika kita ingin melihat sebaran distribusi dari kolom `MonthlyIncome` disetiap departemen?

In [67]:
# visualisasi
px.histogram(attrition,
            x='MonthlyIncome',
            nbins=30,
            color='Department',)

**Pertanyaan Nomor 6**\
6. Bagaimana jika kita ingin membuat sebuah **boxplot** dengan cara memisahkan distribusi dari kolom `MonthlyIncome` berdasarkan departemen dan attrition? 

In [68]:
# visualize
px.box(attrition,
      x = 'MonthlyIncome',
      y = 'Department',
      color = 'Attrition')

**Pertanyaan Nomor 7**\
7. Bagaimana jika kita ingin memvisualisasikan apakah ada korelasi antara kolom `Age` dengan kolom `TotalWorkingYears` pada departemen **Research & Development** ?

In [69]:
# subset the data only Research & Development
rnd = attrition[attrition['Department'] == 'Research & Development']

px.scatter(rnd,
          x = 'Age',
          y = 'TotalWorkingYears',
          facet_col='Attrition')

**Pertanyaan Nomor 8**\
8. Bagaimana jika kita ingin membuat sebuah line chart yang menampilkan antara `Tahun_Masuk` diatas **2010** dengan  `EmployeeNumber` ? 

In [70]:
# buat kolom baru dengan lama tahun awal masuk kerja
attrition['Tahun_Masuk'] = 2022 - attrition['TotalWorkingYears']
attrition.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome      Education EducationField  EmployeeCount  \
0                 1        College  Life Sciences              1   
1                 8  Below College  Life Sciences              1   
2                 2        College          Other              1   
3                 3         Master  Life Sciences              1   
4                 2  Below College        Medical              1   

   EmployeeNumber  ... StandardHours StockOptionLevel  TotalWorkingYears  \
0               1  ...            80                0                  8   
1               2  ...            80                1                 10   
2               4  ...            80                0                  7   
3               5  ...            80                0                  8   
4               7  ...            80                1                  6   

  TrainingTimesLastYear  WorkLifeBalance YearsAtCompany YearsInCurrentRole  \
0                     0              Bad              6                  4   
1                     3           Better             10                  7   
2                     3           Better              0                  0   
3                     3           Better              8                  7   
4                     3           Better              2                  2   

  YearsSinceLastPromotion  YearsWithCurrManager  Tahun_Masuk  
0                       0                     5         2014  
1                       1                     7         2012  
2                       0                     0         2015  
3                       3                     0         2014  
4                       2                     2         2016  

[5 rows x 36 columns]

In [71]:
attrition[attrition['Tahun_Masuk'] >= 2010]

Age Attrition     BusinessTravel  DailyRate              Department  \
0      41       Yes      Travel_Rarely       1102                   Sales   
1      49        No  Travel_Frequently        279  Research & Development   
2      37       Yes      Travel_Rarely       1373  Research & Development   
3      33        No  Travel_Frequently       1392  Research & Development   
4      27        No      Travel_Rarely        591  Research & Development   
...   ...       ...                ...        ...                     ...   
1463   31        No         Non-Travel        325  Research & Development   
1464   26        No      Travel_Rarely       1167                   Sales   
1466   39        No      Travel_Rarely        613  Research & Development   
1467   27        No      Travel_Rarely        155  Research & Development   
1469   34        No      Travel_Rarely        628  Research & Development   

      DistanceFromHome      Education EducationField  EmployeeCount  \
0                    1        College  Life Sciences              1   
1                    8  Below College  Life Sciences              1   
2                    2        College          Other              1   
3                    3         Master  Life Sciences              1   
4                    2  Below College        Medical              1   
...                ...            ...            ...            ...   
1463                 5       Bachelor        Medical              1   
1464                 5       Bachelor          Other              1   
1466                 6  Below College        Medical              1   
1467                 4       Bachelor  Life Sciences              1   
1469                 8       Bachelor        Medical              1   

      EmployeeNumber  ... StandardHours StockOptionLevel  TotalWorkingYears  \
0                  1  ...            80                0                  8   
1                  2  ...            80                1                 10   
2                  4  ...            80                0                  7   
3                  5  ...            80                0                  8   
4                  7  ...            80                1                  6   
...              ...  ...           ...              ...                ...   
1463            2057  ...            80                0                 10   
1464            2060  ...            80                0                  5   
1466            2062  ...            80                1                  9   
1467            2064  ...            80                1                  6   
1469            2068  ...            80                0                  6   

     TrainingTimesLastYear  WorkLifeBalance YearsAtCompany YearsInCurrentRole  \
0                        0              Bad              6                  4   
1                        3           Better             10                  7   
2                        3           Better              0                  0   
3                        3           Better              8                  7   
4                        3           Better              2                  2   
...                    ...              ...            ...                ...   
1463                     2           Better              9                  4   
1464                     2           Better              4                  2   
1466                     5           Better              7                  7   
1467                     0           Better              6                  2   
1469                     3             Best              4                  3   

     YearsSinceLastPromotion  YearsWithCurrManager  Tahun_Masuk  
0                          0                     5         2014  
1                          1                     7         2012  
2                          0                     0         2015  
3                          3              

In [72]:
# Visualize lihat data pegawai yang masuk >= 2010berdasarkan EmployeeNumber nya
# Data wrangling
year_2010 = attrition[attrition['Tahun_Masuk'] >= 2010]
year_2010_agg = year_2010.groupby(['Tahun_Masuk']).count()[['EmployeeNumber']].reset_index()
year_2010_agg


# Visualize
px.line(year_2010_agg,
       x = 'Tahun_Masuk',
       y = 'EmployeeNumber')